sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -1732,44 +1801,348 @@ block_store(T *ptr, size_t byte_offset, simd_view<Toffset, RegionTy> vals,
+    detail::is_device_accessor_with_v<AccessorT,
+                                      detail::accessor_mode_cap::can_write> &&
+    !sycl::detail::acc_properties::is_local_accessor_v<AccessorT>>
+block_store(AccessorT acc, detail::DeviceAccessorOffsetT byte_offset,
+            simd<T, N> vals, PropertyListT props = {}) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
-  block_store<Tx, N>(__ESIMD_DNS::accessorToPointer<Tx>(acc, offset), vals);
+  block_store<T, N>(detail::accessorToPointer<T>(acc, byte_offset), vals,
+                    props);
 #else
-  constexpr unsigned Sz = sizeof(T) * N;
-  static_assert(Sz >= detail::OperandSize::OWORD,
-                "block size must be at least 1 oword");
-  static_assert(Sz % detail::OperandSize::OWORD == 0,
-                "block size must be whole number of owords");
-  static_assert(detail::isPowerOf2(Sz / detail::OperandSize::OWORD),
-                "block must be 1, 2, 4 or 8 owords long");
-  static_assert(Sz <= 8 * detail::OperandSize::OWORD,
-                "block size must be at most 8 owords");
+  constexpr auto L1Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L1_key>(
+          cache_hint::none);
+  constexpr auto L2Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L2_key>(
+          cache_hint::none);
+  static_assert(!PropertyListT::template has_property<cache_hint_L3_key>(),
+                "L3 cache hint is reserved. The old/experimental L3 LSC cache "
+                "hint is cache_level::L2 now.");
+  if constexpr (L1Hint != cache_hint::none || L2Hint != cache_hint::none) {
+    detail::check_cache_hint<detail::cache_action::store, L1Hint, L2Hint>();
+    constexpr int DefaultAlignment = (sizeof(T) <= 4) ? 4 : sizeof(T);
+    constexpr size_t Alignment =
+        detail::getPropertyValue<PropertyListT, alignment_key>(
+            DefaultAlignment);
 
-  auto surf_ind = __esimd_get_surface_index(
-      detail::AccessorPrivateProxy::getQualifiedPtrOrImageObj(acc));
-  __esimd_oword_st<T, N>(surf_ind, offset >> 4, vals.data());
+    simd_mask<1> Mask = 1;
+    detail::block_store_impl<T, N, L1Hint, L2Hint>(
+        acc, byte_offset, vals, Mask, overaligned_tag<Alignment>{});
+  } else {
+    using Tx = detail::__raw_t<T>;
+    constexpr unsigned Sz = sizeof(Tx) * N;
+    static_assert(Sz >= detail::OperandSize::OWORD,
+                  "block size must be at least 1 oword");
+    static_assert(Sz % detail::OperandSize::OWORD == 0,
+                  "block size must be whole number of owords");
+    static_assert(detail::isPowerOf2(Sz / detail::OperandSize::OWORD),
+                  "block must be 1, 2, 4 or 8 owords long");
+    static_assert(Sz <= 8 * detail::OperandSize::OWORD,
+                  "block size must be at most 8 owords");
+
+    auto surf_ind = __esimd_get_surface_index(
+        detail::AccessorPrivateProxy::getQualifiedPtrOrImageObj(acc));
+    __esimd_oword_st<Tx, N>(surf_ind, byte_offset >> 4, vals.data());
```

Comment:
block_store(acc) has some difference comparing to block_load(acc).
The load may be lowered to __esimd_oword_ld, or __esimd_oword_ld_unaligned, while store can only be stored using aligned store.

This code expects from user that `byte_offset` is 16-byte aligned as it was previously (it was not explicitly told in the old comment section).

This place is a bit tricky.
We can choose to say that if (!(b) && (c)) then the function assumes the default alignment is 16 if no alignment props passed. Also, if 'alignment' property 4 or 8 is passed, then this function generates code requiring DG2 or PVC device.

Reply:
Thanks for the in-depth review. I updated the function and doc to expect 16-byte alignment if the property is not passed and !b && C, and if alignment of < 16 is passed we require LSC and DG2, hopefully I did it correctly, please take a look